[PT-D][Tensor parallelism] Add documentations for TP

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #94421


This is far from completed and we will definitely polish it down the road.

Commit:
[PT-D][Tensor parallelism] Add documentations for TP

[ghstack-poisoned]

Commit:
Update on "[PT-D][Tensor parallelism] Add documentations for TP"




[ghstack-poisoned]

wz337 commented on docs/source/distributed.tensor.parallel.rst:
```diff
@@ -1,7 +1,59 @@
+======================================================
+
+We built Tensor Parallelism(TP) on top of DistributedTensor(DTensor) and
+provide several Parallelis styles: Rowwise, Colwise and Pairwise Parallelism.
```
nit: several Parallelism?

fduwjj replied:
opps. Thanks!

wz337 reviewed (approved):
LGTM

wanchaol commented on docs/source/distributed.tensor.parallel.rst:
```diff
@@ -1,7 +1,59 @@
+.. autofunction::  make_output_shard_1d
+
+Currently, there are some constraints which makes it hard for attention module
+to work out of box for TP, so we built this multihead_attention module for TP:
```
We should mention that `parallelize_module` will automatically swap nn.MultiheadAttention to this custom module when specifying PairwiseParallel. We don't want user to surprise

wanchaol commented on docs/source/distributed.tensor.parallel.rst:
```diff
@@ -1,7 +1,59 @@
+
+.. currentmodule:: torch.distributed.tensor.parallel.style
+
+.. autofunction::  make_input_replicate_1d
```
Actually I'm wondering if we make those APIs be something like `mark_input_replicate_1d` instead of `make`? The real thing we are doing is to mark the inputs/outputs instead of turning them into (i.e. we are using `from_local` instead of `distribute_tensor`)

fduwjj replied:
I understand where you are from... Although internally we don't have drastical changes here. From users' perspective we still change a Tensor to DTensor. So I really don't like the word "mark". (Definition: https://www.merriam-webster.com/dictionary/mark). It does not contain any meaning related to change. I would prefer verbs like "change", "convert", "transform", etc.. Or even construct maybe..

wanchaol commented on docs/source/distributed.tensor.parallel.rst:
```diff
@@ -1,7 +1,59 @@
+
+.. autofunction::  make_input_replicate_1d
+.. autofunction::  make_input_shard_1d
+.. autofunction::  make_input_shard_1d_dim_last
```
nit: maybe call it last_dim instead of dim_last?

fduwjj replied:
ok, not a strong opinion here.

wanchaol commented on docs/source/distributed.tensor.parallel.rst:
```diff
@@ -1,7 +1,59 @@
+We built Tensor Parallelism(TP) on top of DistributedTensor(DTensor) and
+provide several Parallelis styles: Rowwise, Colwise and Pairwise Parallelism.
+
+The entrypoint to parallelize your module and use tensor parallelism is:
```
```suggestion
The entrypoint to parallelize your module using Tensor Parallelism is:
```

wanchaol commented on docs/source/distributed.tensor.parallel.rst:
```diff
@@ -1,7 +1,59 @@
+.. autoclass:: torch.distributed.tensor.parallel.style.ParallelStyle
+  :members:
+
+We have following parallel styles:
```
```suggestion
Tensor Parallelism supports the following parallel styles:
```

wanchaol commented on docs/source/distributed.tensor.parallel.rst:
```diff
@@ -1,7 +1,59 @@
+.. autoclass:: torch.distributed.tensor.parallel.style.PairwiseParallel
+  :members:
+
+Because we use DTensors within the TP module, we need to specify how we want
```
```suggestion
Because we use DTensor within Tensor Parallelism, we need to mark the input and output of the module with DTensors.
```

wanchaol commented on docs/source/distributed.tensor.parallel.rst:
```diff
@@ -1,7 +1,59 @@
+.. autofunction::  make_output_tensor
+.. autofunction::  make_output_shard_1d
+
+Currently, there are some constraints which makes it hard for attention module
```
```suggestion
Currently, there are some constraints which makes it hard for the `nn.MultiheadAttention` module work out of box for Tensor Parallelism
```

wanchaol reviewed (commented):
first pass, let's add the experimental line as this is prototype release.

fduwjj commented on docs/source/distributed.tensor.parallel.rst:
```diff
@@ -1,7 +1,59 @@
+
+.. autofunction::  parallelize_module
+
+All parallel styles inherits the following:
```
Got it, sure.

fduwjj commented on docs/source/distributed.tensor.parallel.rst:
```diff
@@ -1,7 +1,59 @@
+
+
+.. currentmodule:: torch.distributed.tensor.parallel.fsdp
+.. autofunction::  is_available
```
Yes, because of 2D hook registration.

fduwjj replied:
Will send a follow-up PR to address the naming of this one.

fduwjj commented on docs/source/distributed.tensor.parallel.rst:
```diff
@@ -1,7 +1,59 @@
+  :members:
+
+We also enabled 2D parallelism to integrate with ``FullyShardedDataParallel``.
+Users just need to call the following API explicitly:
```
The registrations is in the `is_available`

Commit:
Update on "[PT-D][Tensor parallelism] Add documentations for TP"



This is far from completed and we will definitely polish it down the road.


[ghstack-poisoned]

fduwjj commented:
@pytorchbot rebase

Commit:
Update on "[PT-D][Tensor parallelism] Add documentations for TP"



This is far from completed and we will definitely polish it down the road.


[ghstack-poisoned]

fduwjj commented:
@pytorchbot merge
